File: source/jugglinglab/prop/ballProp.java
// ballProp.java
//
// Copyright 2019 by Jack Boyce (jboyce@gmail.com)

package jugglinglab.prop;

import java.util.*;
import java.awt.*;
import java.awt.image.*;
import java.lang.reflect.*;
import java.text.MessageFormat;

import jugglinglab.core.*;
import jugglinglab.util.*;
import jugglinglab.renderer.*;


public class ballProp extends Prop {
    static String[] colornames = {"black", "blue", "cyan", "gray",
        "green", "magenta", "orange", "pink", "red", "white", "yellow"};
    static Color[] colorvals = {Color.black, Color.blue, Color.cyan,
        Color.gray, Color.green, Color.magenta, Color.orange,
        Color.pink, Color.red, Color.white, Color.yellow};

    protected static final int colornum_def = 8;    // red
    protected static final double diam_def = 10.0;  // in cm
    protected static final boolean highlight_def = false;

    protected double        diam = diam_def;    // diameter, in cm
    protected int           colornum = colornum_def;
    protected Color         color;
    protected boolean       highlight = highlight_def;
    // protected int    ball_pixel_size;

    protected BufferedImage ballimage;
    protected double        lastzoom = 0.0;
    // protected int        offsetx, offsety;
    protected Dimension     size;
    protected Dimension     center;
    protected Dimension     grip;
    protected Coordinate    propmax;
    protected Coordinate    propmin;

    @Override
    public String getName() {
        return "Ball";
    }

    @Override
    public Color getEditorColor() {
        return color;
    }

    @Override
    public ParameterDescriptor[] getParameterDescriptors() {
        ParameterDescriptor[] result = new ParameterDescriptor[3];

        ArrayList<String> range = new ArrayList<String>();
        for (int i = 0; i < colornames.length; i++)
            range.add(colornames[i]);

        result[0] = new ParameterDescriptor("color", ParameterDescriptor.TYPE_CHOICE,
                                            range, colornames[colornum_def], colornames[colornum]);
        result[1] = new ParameterDescriptor("diam", ParameterDescriptor.TYPE_FLOAT,
                                            null, new Double(diam_def), new Double(diam));
        result[2] = new ParameterDescriptor("highlight", ParameterDescriptor.TYPE_BOOLEAN,
                                            null, new Boolean(highlight_def), new Boolean(highlight));

        return result;
    }

    @Override
    protected void init(String st) throws JuggleExceptionUser {
        color = Color.red;

        if (st == null) return;
        ParameterList pl = new ParameterList(st);

        String colorstr = pl.getParameter("color");
        if (colorstr != null) {
            Color temp = null;
            if (colorstr.indexOf((int)',') == -1) { // color name
                for (int i = 0; i < colornames.length; i++) {
                    if (colornames[i].equalsIgnoreCase(colorstr)) {
                        temp = colorvals[i];
                        colornum = i;
                        break;
                    }
                }
            } else {    // RGB triplet
                     // delete the '{' and '}' characters first
                String str = colorstr;
                int pos;
                while ((pos = str.indexOf('{')) >= 0) {
                    str = str.substring(0,pos) + str.substring(pos+1,str.length());
                }
                while ((pos = str.indexOf('}')) >= 0) {
                    str = str.substring(0,pos) + str.substring(pos+1,str.length());
                }
                StringTokenizer st2 = new StringTokenizer(str, ",", false);
                if (st2.countTokens() == 3) {
                    int red = 0, green = 0, blue = 0;
                    String token = null;
                    try {
                        token = st2.nextToken().trim();
                        red = Integer.valueOf(token).intValue();
                        token = st2.nextToken().trim();
                        green = Integer.valueOf(token).intValue();
                        token = st2.nextToken().trim();
                        blue = Integer.valueOf(token).intValue();
                    } catch (NumberFormatException nfe) {
                        String template = errorstrings.getString("Error_number_format");
                        Object[] arguments = { token };
                        throw new JuggleExceptionUser("Ball prop color: " + MessageFormat.format(template, arguments));
                    }
                    temp = new Color(red, green, blue);
                } else
                    throw new JuggleExceptionUser("Ball prop color: " + errorstrings.getString("Error_token_count"));
            }

            if (temp != null)
                color = temp;
            else {
                String template = errorstrings.getString("Error_prop_color");
                Object[] arguments = { colorstr };
                throw new JuggleExceptionUser(MessageFormat.format(template, arguments));
            }
        }

        String diamstr = pl.getParameter("diam");
        if (diamstr != null) {
            try {
                Double ddiam = Double.valueOf(diamstr.trim());
                double temp = ddiam.doubleValue();
                if (temp > 0.0)
                    diam = temp;
                else
                    throw new JuggleExceptionUser(errorstrings.getString("Error_prop_diameter"));
            } catch (NumberFormatException nfe) {
                String template = errorstrings.getString("Error_number_format");
                Object[] arguments = { "diam" };
                throw new JuggleExceptionUser("Ball prop diameter: " + MessageFormat.format(template, arguments));
            }
        }

        String highlightstr = pl.getParameter("highlight");
        if (highlightstr != null) {
            Boolean bhighlight = Boolean.valueOf(highlightstr);
            highlight = bhighlight.booleanValue();
        }
    }

    @Override
    public Coordinate getMax() {
        if (this.propmax == null)
            this.propmax = new Coordinate(diam / 2.0, 0.0, diam / 2.0);
        return this.propmax;
    }

    @Override
    public Coordinate getMin() {
        if (this.propmin == null)
            this.propmin = new Coordinate(-diam / 2.0, 0, -diam / 2.0);
        return this.propmin;
    }

    @Override
    public Image getProp2DImage(double zoom, double[] camangle) {
        if (ballimage == null || zoom != lastzoom)  // first call or display resized?
            recalc2D(zoom);
        return ballimage;
    }

    @Override
    public Dimension getProp2DSize(double zoom) {
        if (size == null || zoom != lastzoom)       // first call or display resized?
            recalc2D(zoom);
        return size;
    }

    @Override
    public Dimension getProp2DCenter(double zoom) {
        if (center == null || zoom != lastzoom)
            recalc2D(zoom);
        return center;
    }

    @Override
    public Dimension getProp2DGrip(double zoom) {
        if (grip == null || zoom != lastzoom)       // first call or display resized?
            recalc2D(zoom);
        return grip;
    }

    protected void recalc2D(double zoom) {
        int ball_pixel_size = (int)(0.5 + zoom * diam);
        if (ball_pixel_size < 1)
            ball_pixel_size = 1;
        int offsetx = -ball_pixel_size / 2;
        int offsety = -ball_pixel_size;

        // Create a ball image of diameter ball_pixel_size, and transparent background

        ballimage = new BufferedImage(ball_pixel_size+1, ball_pixel_size+1, BufferedImage.TYPE_INT_ARGB_PRE);
        Graphics2D ballg = ballimage.createGraphics();

        /*
        ballg.setRenderingHint(RenderingHints.KEY_ANTIALIASING,
                               RenderingHints.VALUE_ANTIALIAS_ON);
        */
        if (this.highlight) {
            float highlightOvals = ball_pixel_size / 1.2f;  // Number of concentric circles to draw.
            float[] rgb = new float[3];
            rgb[0] = (float)color.getRed() / 255f;
            rgb[1] = (float)color.getGreen() / 255f;
            rgb[2] = (float)color.getBlue() / 255f;
            // Make the color a little darker so that there is some contrast.
            for (int i = 0; i < rgb.length; i++) {
                rgb[i] = rgb[i] / 2.5f;
            }

            ballg.setColor(new Color(rgb[0], rgb[1], rgb[2]));
            ballg.fillOval(0, 0, ball_pixel_size, ball_pixel_size); // Full sized ellipse.

            // Now draw the highlight on the ball.
            for (int i = 0; i < highlightOvals; i++) {
                // Calculate the new color
                for (int j = 0; j < rgb.length; j++) {
                    rgb[j] = Math.min(rgb[j] + (1f / highlightOvals), 1f);
                }
                ballg.setColor(new Color(rgb[0], rgb[1], rgb[2]));
                ballg.fillOval((int)(i/1.1), (int)(i/2.5),  // Literals control how fast highlight
                                                            // moves right and down respectively.
                               ball_pixel_size - (int)(i*1.3),   // These control how fast the
                               ball_pixel_size - (int)(i*1.3));  // highlight converges to a point.
            }
        } else {
            ballg.setColor(color);
            ballg.fillOval(0, 0, ball_pixel_size, ball_pixel_size);
        }

        size = new Dimension(ball_pixel_size, ball_pixel_size);
        center = new Dimension(ball_pixel_size/2, ball_pixel_size/2);
        grip = new Dimension(ball_pixel_size/2, ball_pixel_size/2);

        lastzoom = zoom;
    }

    /*
    public Object getPropIDX3D() {
        Object result = null;
        try {
            Class ob = Class.forName("idx3d.idx3d_Object");
            Class of = Class.forName("idx3d.idx3d_ObjectFactory");
            Class mat = Class.forName("idx3d.idx3d_Material");
            Method sphere = of.getMethod("SPHERE", new Class[] {Float.TYPE, Integer.TYPE});
            result = sphere.invoke(null, new Object[] {new Float((float)diam/2),
                new Integer(10)});
            Method setcolor = mat.getMethod("setColor", new Class[] {Integer.TYPE});
            Object surf = mat.newInstance();
            setcolor.invoke(surf, new Object[] {new Integer(color.getRGB())});
            Method setmaterial = ob.getMethod("setMaterial", new Class[] {mat});
            setmaterial.invoke(result, new Object[] {surf});
        } catch (ClassNotFoundException e) {
            return null;
        } catch (NoSuchMethodException e) {
            return null;
        } catch (SecurityException e) {
            return null;
        } catch (IllegalAccessException e) {
            return null;
        } catch (InstantiationException e) {
            return null;
        } catch (InvocationTargetException e) {
            return null;
        }
        return result;
    }

    public Coordinate getPropIDX3DGrip() {
        return new Coordinate(0.0, 0.0, -diam/2);       // bottom of ball
    }
    */
}

File: src/jvmMain/kotlin/jugglinglab/prop/BallProp.kt
//
// BallProp.java
//
// Copyright 2002-2025 Jack Boyce and the Juggling Lab contributors
//

package jugglinglab.prop

import jugglinglab.generated.resources.*
import jugglinglab.util.Coordinate
import jugglinglab.util.JuggleExceptionUser
import jugglinglab.util.ParameterDescriptor
import jugglinglab.util.ParameterList
import jugglinglab.util.NumberFormatter.jlParseFiniteDouble
import jugglinglab.util.getStringResource
import java.awt.Color
import java.awt.Dimension
import java.awt.Image
import java.awt.image.BufferedImage
import kotlin.math.max
import kotlin.math.min

class BallProp : Prop() {
    private var diam: Double = DIAM_DEF // diameter, in cm
    private var color: Color = COLOR_DEF
    private var colornum: Int = COLORNUM_DEF
    private var highlight: Boolean = HIGHLIGHT_DEF

    private var ballimage: BufferedImage? = null
    private var lastzoom: Double = 0.0
    private var size: Dimension? = null
    private var center: Dimension? = null
    private var grip: Dimension? = null

    override val type = "Ball"

    override val isColorable = true

    override fun getEditorColor(): Color {
        return color
    }

    override fun getParameterDescriptors(): List<ParameterDescriptor> {
        return listOf(
            ParameterDescriptor(
                "color",
                ParameterDescriptor.TYPE_CHOICE,
                COLOR_NAMES,
                COLOR_NAMES[COLORNUM_DEF],
                COLOR_NAMES[colornum]
            ),
            ParameterDescriptor(
                "diam",
                ParameterDescriptor.TYPE_FLOAT,
                null,
                DIAM_DEF,
                diam
            ),
            ParameterDescriptor(
                "highlight",
                ParameterDescriptor.TYPE_BOOLEAN,
                null,
                HIGHLIGHT_DEF,
                highlight
            ),
        )
    }

    @Throws(JuggleExceptionUser::class)
    override fun init(st: String?) {
        if (st == null) {
            return
        }
        val pl = ParameterList(st)

        val colorstr = pl.getParameter("color")
        if (colorstr != null) {
            var temp: Color? = null

            if (colorstr.indexOf(',') == -1) {  // color name
                for (i in COLOR_NAMES.indices) {
                    if (COLOR_NAMES[i].equals(colorstr, ignoreCase = true)) {
                        temp = COLOR_VALS[i]
                        colornum = i
                        break
                    }
                }
            } else {  // RGB or RGBA
                val tokens = colorstr.replace("{", "").replace("}", "").split(',')
                if (tokens.size == 3 || tokens.size == 4) {
                    try {
                        val intTokens = tokens.map { it.trim().toInt() }
                        val red = intTokens[0]
                        val green = intTokens[1]
                        val blue = intTokens[2]
                        val alpha = if (tokens.size == 4) intTokens[3] else 255

                        if (listOf(red, green, blue, alpha).any { it !in 0..255 }) {
                            val message = getStringResource(Res.string.error_prop_color, colorstr)
                            throw JuggleExceptionUser(message)
                        }

                        temp = Color(red, green, blue, alpha)
                    } catch (_: NumberFormatException) {
                        val message = getStringResource(Res.string.error_prop_color, colorstr)
                        throw JuggleExceptionUser(message)
                    }
                } else {
                    val message = getStringResource(Res.string.error_token_count)
                    throw JuggleExceptionUser(message)
                }
            }

            if (temp != null) {
                color = temp
            } else {
                val message = getStringResource(Res.string.error_prop_color, colorstr)
                throw JuggleExceptionUser(message)
            }
        }

        val diamstr = pl.getParameter("diam")
        if (diamstr != null) {
            try {
                val temp = jlParseFiniteDouble(diamstr.trim { it <= ' ' })
                if (temp > 0) {
                    diam = temp
                } else {
                    val message = getStringResource(Res.string.error_prop_diameter)
                    throw JuggleExceptionUser(message)
                }
            } catch (_: NumberFormatException) {
                val message = getStringResource(Res.string.error_number_format, "diam")
                throw JuggleExceptionUser(message)
            }
        }

        val highlightstr = pl.getParameter("highlight")
        if (highlightstr != null) {
            highlight = highlightstr.toBoolean()
        }
    }

    override fun getMax(): Coordinate {
        return Coordinate(diam / 2, 0.0, diam / 2)
    }

    override fun getMin(): Coordinate {
        return Coordinate(-diam / 2, 0.0, -diam / 2)
    }

    override fun getWidth(): Double {
        return diam
    }

    override fun getProp2DImage(zoom: Double, camangle: DoubleArray): Image? {
        if (ballimage == null || zoom != lastzoom) {
            // first call or display resized?
            recalc2D(zoom)
        }
        return ballimage
    }

    override fun getProp2DSize(zoom: Double): Dimension? {
        if (size == null || zoom != lastzoom) {
            // first call or display resized?
            recalc2D(zoom)
        }
        return size
    }

    override fun getProp2DCenter(zoom: Double): Dimension? {
        if (center == null || zoom != lastzoom) {
            recalc2D(zoom)
        }
        return center
    }

    override fun getProp2DGrip(zoom: Double): Dimension? {
        if (grip == null || zoom != lastzoom) {
            // first call or display resized?
            recalc2D(zoom)
        }
        return grip
    }

    private fun recalc2D(zoom: Double) {
        var ballPixelSize = (0.5 + zoom * diam).toInt()
        ballPixelSize = max(ballPixelSize, 1)

        // create a ball image of diameter ball_pixel_size, and transparent background
        ballimage = BufferedImage(
            ballPixelSize + 1, ballPixelSize + 1, BufferedImage.TYPE_INT_ARGB_PRE
        )
        val ballg = ballimage!!.createGraphics()

        /*
        ballg.setRenderingHint(RenderingHints.KEY_ANTIALIASING,
                               RenderingHints.VALUE_ANTIALIAS_ON);
        */
        if (highlight) {
            val highlightOvals = ballPixelSize / 1.2f // Number of concentric circles to draw.
            val rgb = FloatArray(4)
            rgb[0] = color.red.toFloat() / 255f
            rgb[1] = color.green.toFloat() / 255f
            rgb[2] = color.blue.toFloat() / 255f
            rgb[3] = color.alpha.toFloat() / 255f

            // make the color a little darker so that there is some contrast
            for (i in 0..2) {
                rgb[i] = rgb[i] / 2.5f
            }

            ballg.color = Color(rgb[0], rgb[1], rgb[2], rgb[3])
            ballg.fillOval(0, 0, ballPixelSize, ballPixelSize) // Full sized ellipse.

            // draw the highlight on the ball
            var i = 0
            while (i < highlightOvals) {
                // Calculate the new color
                for (j in 0..2) {
                    rgb[j] = min(rgb[j] + (1f / highlightOvals), 1f)
                }
                ballg.color = Color(rgb[0], rgb[1], rgb[2], rgb[3])
                ballg.fillOval(
                    (i / 1.1).toInt(),
                    (i / 2.5).toInt(),  // literals control how fast highlight
                    // moves right and down respectively.
                    ballPixelSize - (i * 1.3).toInt(),  // these control how fast the
                    ballPixelSize - (i * 1.3).toInt()
                ) // highlight converges to a point.
                i++
            }
        } else {
            ballg.color = color
            ballg.fillOval(0, 0, ballPixelSize, ballPixelSize)
        }

        size = Dimension(ballPixelSize, ballPixelSize)
        center = Dimension(ballPixelSize / 2, ballPixelSize / 2)
        grip = Dimension(ballPixelSize / 2, ballPixelSize / 2)

        lastzoom = zoom
    }

    companion object {
        private val COLOR_DEF: Color = Color.red
        private const val COLORNUM_DEF: Int = 9  // red
        private const val DIAM_DEF: Double = 10.0  // in cm
        private const val HIGHLIGHT_DEF: Boolean = false
    }
}
